File: matlab/unit/demo_ar.m
%  Running this demos estimation of AR model structure 
%  using prediction error method

clear; close all;  
global dm; if isempty(dm), clear global dm; dm=0; end
global dsp; if isempty(dsp), clear global dsp; dsp=1; end
global trans;

if dsp>1, clc; echo on; end
OPT.dsp = dsp;

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%
%  Specify Experiment Conditions                    
%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

T   = 1;         % Sampling Period in seconds
N   = 500;       % Number of Samples
var = 1e-1;      % White innovations input variance

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%
%  Specify Noise Colouring
%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

aq = real(poly([0.5,0.7,0.9,0.9*exp(j*pi/6),0.9*exp(-j*pi/6)]));  

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%
%  Simulate a data record                    
%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

noise = sqrt(var)*randn(1,N);
y     = filter(1,aq,noise(:));
Z.y   = y(:);

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%
%  Specify Model Structure                    
%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

Mq.nA = length(aq)-1; 
Mq.T  = T; 
Mq.type = 'ar';

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%
%  Estimate model on basis of observed data
%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

Gq = est(Z,Mq,OPT); 

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%
%  Plot the results
%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

if dsp, 
	Gt.A = aq;
	Gt.T = T;
 Gt.w = Gq.w; 
	Gt.type = 'ar';
	Gt.disp.colour='b';
	Gt.disp.legend = 'True Spectral Factor';
	showbode(Gt,Gq);
end

echo off;

if dm
 disp('  ')
 disp('---------------------------------------------------------------------')
 disp('  ')
 disp('You now have access to the MATLAB workspace so that you may examine')
 disp('the results of this simulation.  To return to the demos, type "dbcont"')
 disp(' ')
 keyboard; 
end;

File: src/uonidtoolbox/demo/_demo_ar.py

import uonidtoolbox as unit
import numpy as np
import scipy


def demo_ar(disp=1):

    OPT = unit.struct()
    OPT.dsp = disp

    # ====================================================
    # Specify Experiment Conditions
    # ====================================================
    T   = 0.5   # Sampling period in seconds
    N   = 500   # Number of Samples
    var = 1e-1  # Measurement Noise Variance


    # ====================================================
    # Specify Noise Colouring
    # ====================================================
    aq = np.real(np.poly([
        0.5, 
        0.7, 
        0.9, 
        0.9*np.exp( 1j*np.pi/6), 
        0.9*np.exp(-1j*np.pi/6)
    ]))


    # ====================================================
    # Simulate a data record
    # ====================================================
    Z       = unit.struct()
    noise   = np.sqrt(var)*np.random.randn(1,N)
    Z.y     = scipy.signal.lfilter(1, aq, noise)


    # ====================================================
    # Specify Model Structures
    # ====================================================
    Mq          = unit.struct()
    Mq.nA       = aq.shape[0]-1
    Mq.T        = T
    Mq.type     = 'ar'


    # ====================================================
    # Estimate on basis of noise corrupted data
    # ====================================================
    Gq = unit.est(Z,Mq,OPT)


    # ====================================================
    # Plot the results
    # ====================================================
    if OPT.dsp:
        Gt              = unit.struct()
        Gt.disp         = unit.struct()
        Gt.A            = aq
        Gt.T            = T
        Gt.w            = Gq.w
        Gt.type         = 'ar'
        # Gt.colour       = 'b'
        Gt.disp.legend  = 'True Spectral Factor'
        
        unit.plotting.showbode([Gt, Gq])
    #endif

#endfunction

